prometheus/graphite/bridge.go

```diff
@@ -0,0 +1,224 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package graphite
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Config defines the graphite bridge config.
+type Config struct {
+	// The url to push data to. Required.
+	URL string
+
+	// The interval to use for pushing data to Graphite. Defaults to 15 seconds.
+	Interval time.Duration
+
+	// The Gatherer to use for metrics. Defaults to prometheus.DefaultGatherer.
+	Gatherer prometheus.Gatherer
+
+	// The prefix for your graphite metric. Defaults to empty string.
+	Prefix string
+}
+
+// Bridge pushes metrics to the configured graphite server.
+type Bridge struct {
+	url      string
+	interval time.Duration
+	prefix   string
+	stopc    chan struct{}
+
+	g prometheus.Gatherer
+}
+
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{
+		stopc: make(chan struct{}),
+	}
+
+	if c.URL == "" {
+		return nil, errors.New("graphite bridge: no url given")
+	}
+	b.url = c.URL
+
+	if c.Gatherer == nil {
+		b.g = prometheus.DefaultGatherer
+	} else {
+		b.g = c.Gatherer
+	}
+
+	if c.Prefix != "" {
+		b.prefix = c.Prefix
+	}
+
+	var z time.Duration
+	if c.Interval == z {
+		b.interval = 15 * time.Second
+	} else {
+		b.interval = c.Interval
+	}
+
+	go b.loop()
+
+	return b, nil
+}
+
+func (b *Bridge) Stop() {
+	b.stopc <- struct{}{}
+}
+
+func (b *Bridge) loop() {
+	ticker := time.NewTicker(b.interval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := b.Push(); err != nil {
+				// TODO: Use the right logger.
+				log.Printf("%v", err)
+			}
+		case <-b.stopc:
+			return
+		}
+	}
+}
+
+func (b *Bridge) Push() error {
+	mfs, err := b.g.Gather()
+	if err != nil {
+		return err
+	}
+
+	now := time.Now().Unix()
+	// TODO: Write directly to conn?
+	buf, err := toReader(mfs, b.prefix, now)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Should we expose a deadline to the user?
+	conn, err := net.Dial("tcp", b.url)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	_, err = io.Copy(conn, buf)
+	return err
+}
+
+// TODO: Do we want to allow partial writes? i.e., Skip a failed metric, but
+// try to write all metrics?
+func toReader(mfs []*dto.MetricFamily, prefix string, now int64) (*bytes.Buffer, error) {
+	// TODO: Snag the buffer pool from promhttp/http.go
+	var (
+		buf bytes.Buffer
+		err error
+	)
+
+	for _, mf := range mfs {
+		for _, m := range mf.GetMetric() {
+			sort.Sort(prometheus.LabelPairSorter(m.GetLabel()))
+
+			parts := []string{prefix, mf.GetName()}
+			for _, lp := range m.GetLabel() {
+				parts = append(parts, sanitize(lp.GetName())+"."+sanitize(lp.GetValue()))
+			}
+
+			switch mf.GetType() {
+			case dto.MetricType_SUMMARY:
```

Comment:
It's probably better to convert this into a model.Vector rather than reimplementing all that logic again (you're not handling the case where the +Inf bucket is missing correctly for example).